### ISSUE TITLE
http2 websockets leaks while using open ssl for ping and pong 

### ISSUE THREAD

**Issue by kerdemdemir** [adamdruppe/arsd#466]
I am listening  ~100 web sockets from binance.com to keep up with live transactions(every single buy and sell order). 

I have been having memory leak problems for a long time and have to restart my application after 3-4 hours. Because I am running out of my 32gb ram. 

Now I want to listening order book data as well(supply and demand book) but my applications just run out of memory after 5 mins.  

I use many different tools to find where is leak. The best was gprefd. I found out there were too many allocations from OpenSSL and Arsd. 

I used gdb and put breakpoints to malloc to confirm that and the stack trace always showed OpenSSLSocket's send and receive. 

I changed the "int pingFrequency = 5000;" to  a very big number and now I can run my application for 6-8 hours without running out of memory. But the memory leak still continues because I just controlled the pings that we send but also my server(binance.com) sends me pong approximately every 3 minutes. And pongs uses the same OpenSSLSocket class and its leaking functions send&recieve. 

I can attach the gdb stack outputs and gpref results if it is requested. 

Thanks for the awesome Arsd library, I removed all vibe.d stuff and replaced it with Arsd :)

**Comment by adamdruppe** [comment]
I might have forgotten to call some openssl cleanup function... post what you have and i can see.

**Comment by adamdruppe** [comment]
huh so it is right inside the openssl functions. i know they keep some pieces around to chunk it right but.... huh

might have to check the docs again and see what im missing

**Comment by kerdemdemir** [comment]
Hi @adamdruppe, 

Claude AI suggested that I fix in http2.d: 

<img width="1522" height="962" alt="Image" src="https://github.com/user-attachments/assets/a941b3bb-b85b-4bc1-aef1-a7e917853608" />

Do you think they might have some value?

**Comment by adamdruppe** [comment]
I don't think that specifically since it calls socket.close when it gets a close event (see line 5485 and a couple other places) so that should be covered.

 There is a function `freeSsl` that is called from the destructor that is supposed to do openssl stuff but maybe the destructor is never running, if the socket is not unregistered, it will never be free in the GC's eyes.

So... i could have missed a spot in unregistering. Freeing that ssl stuff prolly shouldn't wait for the GC anyway, once it is closed and unregistered it could be done immediately i believe. Or an exception being thrown in an onclose handler would cancel that unregister, that could be it too.


Do you open a bunch of connections or just one?

**Comment by adamdruppe** [comment]
there's a receive buffer that grows if messages are large, do you have any big individual messages too?

i wonder if i can reproduce this program locally here

**Comment by adamdruppe** [comment]
taking another look at the code here, each websocket send (including pings and pongs) do allocate small arrays, but the GC should be picking them up quickly. SSL_write doesn't say anything about having to free things there, it is supposed to be auto managed.

I have this discord bot that uses websocket client and runs a long time.... memory appears stable after startup.

is there any way you can post a full program that demonstrates the problem?

**Comment by adamdruppe** [comment]
my one bit went from 12 MB to 18 MB in the first couple minutes. Since then over the last 5 hours it has crept up to about 22 MB. Interesting...

**Comment by adamdruppe** [comment]
one idea would be if it is a bunch of new websockets being created and reconnected but not generating enough trash to trigger the GC of the old ones, maybe that's keep some openssl junk around. just wish i could produce the problem reliably in minutes lol. but trying to call freeSsl in the close method to see what happens. the memory pattern doesn't look  much different...

**Comment by adamdruppe** [comment]
After overnight it is still at just a bit over 22 MB. so maybe this helped? hard to say for sure.

**Comment by adamdruppe** [comment]
https://github.com/adamdruppe/arsd/commit/97e560bcffd97a4bd39ebc0ba836624f7e1a2c7a

this commit might help, the freeSsl call in the close() function has given more stability to my one bot program. Still a little over 22 MB two days later. You might try it too.

**Comment by kerdemdemir** [comment]
Currently only ArsdMultiThreadSocket is in use but I tried in past ArsdPollingSocket as well. It was like 1-2 years ago I don't remember the exact crash but if you are interested I can also try again and start another discussion about it.
